channeldb/invoices.go

```diff
@@ -439,51 +450,58 @@ func (d *DB) QueryInvoices(q InvoiceQuery) (InvoiceSlice, error) {
-	// records we need to skip. We'll also keep track of the record offset
-	// as that's part of the final return value.
-	invoicesToSkip := q.IndexOffset
-	invoiceOffset := q.IndexOffset
-
 	err := d.View(func(tx *bolt.Tx) error {
 		// If the bucket wasn't found, then there aren't any invoices
 		// within the database yet, so we can simply exit.
 		invoices := tx.Bucket(invoiceBucket)
 		if invoices == nil {
 			return ErrNoInvoicesCreated
 		}
-		invoiceAddedIndex := invoices.Bucket(addIndexBucket)
-		if invoiceAddedIndex == nil {
+		invoiceAddIndex := invoices.Bucket(addIndexBucket)
+		if invoiceAddIndex == nil {
 			return ErrNoInvoicesCreated
 		}
 
-		// We'll be using a cursor to seek into the database, so we'll
-		// populate byte slices that represent the start of the key
-		// space we're interested in.
-		var startIndex [8]byte
-		switch q.PendingOnly {
-		case true:
-			// We have to start from the beginning so we know
-			// how many pending invoices we're skipping.
-			byteOrder.PutUint64(startIndex[:], uint64(1))
-		default:
-			// We can seek right to the invoice offset we want
-			// to start with.
-			invoicesToSkip = 0
-			byteOrder.PutUint64(startIndex[:], uint64(invoiceOffset+1))
+		// We'll be using a cursor to seek into the database and return
+		// a slice of invoices. We'll need to determine where to start
+		// our cursor depending on the parameters set within the query.
+		cursor := invoiceAddIndex.Cursor()
+		var invoiceKey []byte
+
+		// It's possible that the index offset is set to 0 due to it
+		// either being explicitly set or not by the user. In order to
+		// handle this edge case, we'll look at the Reversed flag to
+		// determine whether we should show the first or last
+		// NumMaxInvoices invoices.
+		if q.IndexOffset == 0 {
+			if q.Reversed {
+				_, invoiceKey = cursor.Last()
+			} else {
+				_, invoiceKey = cursor.First()
+			}
+		} else {
+			// Otherwise, we'll seek to the given offset and begin
+			// iterating either forwards or backwards depending on
+			// the Reversed flag.
+			var startIndex [8]byte
+			if q.Reversed {
+				byteOrder.PutUint64(startIndex[:], q.IndexOffset-1)
+			} else {
+				byteOrder.PutUint64(startIndex[:], q.IndexOffset+1)
+			}
+			_, invoiceKey = cursor.Seek(startIndex[:])
 		}
 
-		// If we know that a set of invoices exists, then we'll begin
-		// our seek through the bucket in order to satisfy the query.
-		// We'll continue until either we reach the end of the range,
-		// or reach our max number of events.
-		cursor := invoiceAddedIndex.Cursor()
-		_, invoiceKey := cursor.Seek(startIndex[:])
-		for ; invoiceKey != nil; _, invoiceKey = cursor.Next() {
+		// forEachInvoice is a helper closure that will determine
+		// whether we should include an invoice within the response to
+		// the query. errDone is returned if we've reached the maximum
+		// number of invoices in the response.
+		errDone := errors.New("done")
+		forEachInvoice := func(invoiceKey []byte) error {
```

Comment:
Would it be possible to (for readability and a smaller diff) to define 
```go
nextKey := func(c Cursor) {
    if q.Reverserd {
       return c.Prev()
    }
    return c.Next()
}

for ; invoiceKey != nil; _, invoiceKey = nextKey(cursor) {
```

Reply:
Good idea! Fixed.